File: 917 - Reverse Only Letters.cpp
#include <cctype>

class Solution {
public:
    string reverseOnlyLetters(string S) {
        int len = S.size();
        int i = 0, j = len - 1;
        while (i < j) {
            if (i < len && !isalpha(S[i])) {
                i++;
            }
            if (j >= 0 && !isalpha(S[j])) {
                j--;
            }
            if (i < j && isalpha(S[i]) && isalpha(S[j])) {
                char t = S[i];
                S[i++] = S[j];
                S[j--] = t;
            }
        }
        return S;
    }
};

File: 917 - Reverse Only Letters.py
class Solution:
    def reverseOnlyLetters(self, S: str) -> str:
        sub = ''
        S = [i for i in S]
        for c in S:
            if c.isalpha():
                sub += c
        sub = list(reversed(sub))
        i = 0
        for idx, c in enumerate(S):
            if c.isalpha():
                S[idx] = sub[i]
                i += 1
        return ''.join(S)